src/toxcore/dht/server.rs

```diff
@@ -109,28 +110,36 @@ impl Server {
     }
 
     /// create new client
-    pub fn create_client(&self, addr: &SocketAddr, pk: PublicKey) -> Client {
-        let precomputed_key = encrypt_precompute(&pk, &self.sk);
-        Client::new(precomputed_key, self.pk, *addr, self.tx.clone())
+    pub fn create_client(&self, addr: &SocketAddr, pk: PublicKey, peers_cache: &HashMap<PublicKey, Client>) -> Client {
+        if let Some(client) = self.get_client(&pk, peers_cache) {
+            client
+        } else {
+            let precomputed_key = encrypt_precompute(&pk, &self.sk);
+            Client::new(precomputed_key, self.pk, *addr, self.tx.clone())
+        }
     }
     /// get client from cache
-    pub fn get_client(&self, pk: &PublicKey) -> Option<Client> {
-        let state = self.state.read();
+    pub fn get_client(&self, pk: &PublicKey, peers_cache: &HashMap<PublicKey, Client>) -> Option<Client> {
         // Client entry is inserted before sending *Request.
-        if let Some(client) = state.peers_cache.get(pk) {
+        if let Some(client) = peers_cache.get(pk) {
             Some(client.clone())
         }
         else {
             None
         }
     }
-    /// send PingRequest to all peer in kbucket    
+    /// remove timed-out clients
+    pub fn remove_timedout_clients(&self, timeout: Duration) -> IoFuture<()> {
+        let mut state = self.state.write();
+        state.peers_cache.retain(|&_pk, ref client|
+            client.last_resp_time.elapsed() <= timeout);
+        Box::new(future::ok(()))
```

Comment:
It seems no need to return future here.

Reply:
Why not just unit?

Reply:
Because later there could be a smart Future Loop that will shutter the time inside `retain` into many pieces.

Reply:
Do you really believe this will happen? :)

Reply:
Yes